### PR TITLE
#13139 collision_strategy=overwrite from query

### PR DESCRIPTION
**Pull request by alrocar**
Add support for `collision_strategy=overwrite` when creating a dataset from a query. See #13139

### CODE REVIEW

File: app/models/data_import.rb [hunk 1]
```diff
@@ -536,6 +545,52 @@ def import_from_query(name, query)
     table_name
   end
 
+  def overwrite_table_from_query(new_table_name, overwrite_table_name, query, user)
```
Comment (juanignaciosl) [inline app/models/data_import.rb]: There's a lot of duplicated code here with `overwrite_register, isn't there? It's hard for me to make a complete proposal in CR, but we could pair for this (or any other team member that is available). Although this is a nice first step, I'd continue refactoring in order to remove the duplication.

File: app/models/data_import.rb [hunk 2]
```diff
@@ -536,6 +545,52 @@ def import_from_query(name, query)
...
+    end
+
+    table_setup.cartodbfy(overwrite_table_name)
+    table_setup.fix_oid(overwrite_table_name)
+    table_setup.run_index_statements(index_statements, database)
+    table_setup.recreate_overviews(overwrite_table_name)
+    table_setup.update_cdb_tablemetadata(overwrite_table_name)
+  end
+
+  def drop(database, table_name)
```
Comment (juanignaciosl) [inline app/models/data_import.rb]: More duplication.

File: spec/models/data_import_spec.rb [hunk 3]
```diff
@@ -116,6 +116,27 @@
     data_import.table_name.should eq 'walmart_latlon'
     data_import.user.in_database["select count(*) from #{data_import.table_name}"].all[0][:count].should eq 2
     user_tables_should_be_registered
+
+    data_import = create_import(overwrite: true, truncated: false, from_query: 'select * from walmart_latlon limit 1')
```
Comment (juanignaciosl) [inline spec/models/data_import_spec.rb]: We'd normally have a separate test case for this issue. Nevertheless, as this test is pretty screwed up already and your addition is nice, just add a comment explaining the case and linking to the issue (one line, no more is needed).

File: app/connectors/importer.rb [hunk 4]
```diff
@@ -265,46 +265,56 @@ def data_import
...
 
-        move_to_schema(result, result.table_name, ORIGIN_SCHEMA, @destination_schema)
-        @table_setup.cartodbfy(result.table_name)
-
         database.transaction do
           log("Replacing #{name} with #{result.table_name}")
           begin
-            drop("\"#{@destination_schema}\".\"#{name}\"")
-            rename(result, result.table_name, name, @destination_schema)
+            if from_query.nil?
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: I'm not sure at all about this, but maybe you can give this a thought. Again, this is two different behaviors depending on a parameter. Maybe, instead of adding the condition, you could pass a block and use `yield` instead, so the caller decides what to do in the transaction. The old `overwrite_register(result,name)` would be replaced by 

```ruby
overwrite_register(result,name) do
  drop("\"#{@destination_schema}\".\"#{name}\"")
  rename(result, result.table_name, name, @destination_schema)
end
```

And the new one would be:

```ruby
overwrite_register(result,name) do
  database.execute(%{CREATE TABLE #{result.table_name} AS #{from_query}})
  drop("\"#{@destination_schema}\".\"#{name}\"")
  database.execute(%{
    ALTER TABLE "#{@destination_schema}"."#{result.table_name}" RENAME TO "#{name}"
  })
end
```

What do you think? As said, I'm not 100% sure.

File: app/models/data_import.rb [hunk 5]
```diff
@@ -520,9 +520,14 @@ def import_from_query(name, query)
     end
 
     table_name = Carto::ValidTableNameProposer.new.propose_valid_table_name(name, taken_names: taken_names)
-    # current_user.db_services.in_database.run(%{CREATE TABLE #{table_name} AS #{query}})
-    current_user.db_service.in_database_direct_connection(statement_timeout: DIRECT_STATEMENT_TIMEOUT) do |user_direct_conn|
+
+    if overwrite_strategy?
+      overwrite_table_from_query(table_name, name, query)
+      results.push CartoDB::Importer2::Result.new(success: true, error: nil)
```
Comment (juanignaciosl) [inline app/models/data_import.rb]: `from_table` already pushes a `Result`. `from_table` calls `import_from_query`. Because of that, I think that two results would be added `if overwrite_strategy?` is true. Am I right? Is that ok?

File: app/models/data_import.rb [hunk 6]
```diff
@@ -536,6 +541,19 @@ def import_from_query(name, query)
     table_name
   end
 
+  def overwrite_table_from_query(new_table_name, overwrite_table_name, query)
+    importer, = new_importer_runner(nil, nil, nil)
```
Comment (juanignaciosl) [inline app/models/data_import.rb]: Have you considered creating two methods instead of one?

File: app/connectors/importer.rb (hunk 4 above)
Comment (alrocar) [reply app/connectors/importer.rb]: LGTM, I'm giving it a try

File: app/models/data_import.rb (hunk 5 above)
Comment (alrocar) [reply app/models/data_import.rb]: I had to add a [return condition](https://github.com/CartoDB/cartodb/pull/13399/files#diff-dc1988508a82de46843e779e617721d8R491) to the from_table method because the rest of the logic of the method does not apply to this case.

File: app/connectors/importer.rb [hunk 7]
```diff
@@ -85,7 +85,12 @@ def register(result)
         overwrite = overwrite_strategy? && taken_names.include?(name)
 
         if overwrite
-          overwrite_register(result,name)
+          raise ::CartoDB::Importer2::IncompatibleSchemas.new unless compatible_schemas_for_overwrite?(name)
+          move_to_schema(result, result.table_name, ORIGIN_SCHEMA, @destination_schema)
+          overwrite_register(result, name) do
```
Comment (amartinj) [inline app/connectors/importer.rb]: If you find a way to reuse this code not using a block, it would be great

File: spec/models/data_import_spec.rb [hunk 8]
```diff
@@ -190,20 +212,37 @@ def user_tables_should_be_registered
...
-      table_name: 'walmart_latlon',
-      state: "pending",
-      success: false,
-      updated_at: Time.now,
-      created_at: Time.now,
-      original_url: Rails.root.join("spec/support/data/walmart_latlon.csv").to_s,
-      cartodbfy_time: 0.0,
-      collision_strategy: overwrite ? 'overwrite' : nil
-    )
+  def create_import(user: @user, overwrite:, truncated:, incomplete_schema: false, from_query: '')
```
Comment (amartinj) [inline spec/models/data_import_spec.rb]: I see two different behaviours based on `from_query`. What about splitting into two methods?

File: spec/models/data_import_spec.rb [hunk 9]
```diff
@@ -116,6 +116,28 @@
     data_import.table_name.should eq 'walmart_latlon'
     data_import.user.in_database["select count(*) from #{data_import.table_name}"].all[0][:count].should eq 2
     user_tables_should_be_registered
+
```
Comment (amartinj) [inline spec/models/data_import_spec.rb]: This test is testing that tables are being overwritten when set to but, instead of adding more use cases, I'd create a new test simply for testing that it works when importing from a query

File: app/connectors/importer.rb (hunk 7 above)
Comment (juanignaciosl) [reply app/connectors/importer.rb]: I suggested the block. Although it must be documented, it's lightweight and readable. What would you suggest?

Comment (amartinj) [reply app/connectors/importer.rb]: Nothing specifically. Just to think it over once again.

File: spec/models/data_import_spec.rb [hunk 10]
```diff
@@ -118,6 +118,31 @@
     user_tables_should_be_registered
   end
 
+  it 'should overwrite dataset with query if collision_strategy is set to overwrite' do
```
Comment (amartinj) [inline spec/models/data_import_spec.rb]: These tests don't pass. Maybe they do if they are run in a specific order among the other tests in this file.

You need to set up an environment and make sure it's as expected, run something, and run some expectations again to check that what you expected actually happened. Read lines below

File: spec/models/data_import_spec.rb [hunk 11]
```diff
@@ -118,6 +118,31 @@
...
+    # overwriting from a sql is a different case needs to be tackled -> https://github.com/CartoDB/cartodb/issues/13139
+    query = 'select * from walmart_latlon limit 1'
```
Comment (amartinj) [inline spec/models/data_import_spec.rb]: In this case, unless some of the other tests are run first, there won't be a wallmart_latlon table.

File: spec/models/data_import_spec.rb [hunk 12]
```diff
@@ -118,6 +118,31 @@
...
+    data_import = create_import_from_query(overwrite: true, from_query: query)
+    data_import.run_import!
+    carto_user.reload
+    carto_user.visualizations.count.should eq 3
```
Comment (amartinj) [inline spec/models/data_import_spec.rb]: It's fine that you assert that the number of visualizations is under control, but in relation to something. You should check it's three before and after running your import (or whatever you are making sure of). You could also make sure that the original table has been overwritten by your import by checking the number of rows like it's done [here](https://github.com/CartoDB/cartodb/blob/0adc84206576f5b472e3cba9bd1ad71aef7ed79d/spec/models/data_import_spec.rb#L99).
Basically: try to make clear what are you testing and don't depend on the state left by other test that (might or not) run prior to this

File: spec/models/data_import_spec.rb (hunk 10 above)
Comment (alrocar) [reply spec/models/data_import_spec.rb]: Upss I just moved the test and forgot to make it work........ Friday...
